### PR TITLE
 strictNullChecks: First batch

### PR DESCRIPTION
**Pull request by tskarhed**
#18389

**Special notes for your reviewer**:
First one of these PRs, so please give tons of feedback!

### CODE REVIEW

File: packages/grafana-data/src/types/navModel.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ export interface NavModelItem {
   hideFromTabs?: boolean;
   hideFromMenu?: boolean;
   divider?: boolean;
-  children?: NavModelItem[];
+  children?: NavModelItem[] | undefined;
```
Comment (torkelo) [inline packages/grafana-data/src/types/navModel.ts]: this looks strange, the? should be the same as it is being undefined

File: public/app/plugins/panel/singlestat/module.ts [hunk 2]
```diff
@@ -182,9 +182,9 @@ class SingleStatCtrl extends MetricsPanelCtrl {
     if (tableData.columns.length === 1) {
       this.panel.tableColumn = tableData.columns[0].text;
     } else {
-      this.panel.tableColumn = _.find(tableData.columns, col => {
+      this.panel.tableColumn = _.find(tableData.columns!, col => {
         return col.type !== 'time';
-      }).text;
+      })!.text;
```
Comment (torkelo) [inline public/app/plugins/panel/singlestat/module.ts]: can columns and text nullable/undefined? that can't be right

Comment (dehrax) [reply public/app/plugins/panel/singlestat/module.ts]: `_.find` can return `undefined`, not sure if it ever will in this case though.

Comment (dehrax) [reply public/app/plugins/panel/singlestat/module.ts]: I added `tableColumn` as an array of MutableColumns and undefined. Would this be a correct approach?

**Comment by torkelo** [comment]
Fixed a couple of issues, this is definitely more tricky than no implicit any!

File: public/app/plugins/panel/table/transformers.ts [hunk 3]
```diff
@@ -191,7 +191,7 @@ transformers['json'] = {
       const maxDocs = Math.min(series.datapoints.length, 100);
       for (let y = 0; y < maxDocs; y++) {
         const doc = series.datapoints[y];
-        const flattened = flatten(doc, null);
+        const flattened = flatten(doc, {});
```
Comment (torkelo) [inline public/app/plugins/panel/table/transformers.ts]: can you modify the flatten object instead, and make the opts param optional?

File: public/app/plugins/panel/table2/types.ts [hunk 4]
```diff
@@ -27,7 +27,7 @@ export const defaults: Options = {
       alias: '',
       decimals: 2,
       colors: ['rgba(245, 54, 54, 0.9)', 'rgba(237, 129, 40, 0.89)', 'rgba(50, 172, 45, 0.97)'],
-      colorMode: null,
+      colorMode: undefined,
```
Comment (torkelo) [inline public/app/plugins/panel/table2/types.ts]: no need to define something undefined

File: public/vendor/ansicolor/ansicolor.ts [hunk 5]
```diff
@@ -60,9 +60,9 @@ const clean = (obj: any) => {
 /*  ------------------------------------------------------------------------ */
 
 class Color {
-  background: boolean;
-  name: string;
-  brightness: number;
+  background: boolean | undefined;
+  name: string | undefined;
+  brightness: number | undefined;
```
Comment (dprokop) [inline public/vendor/ansicolor/ansicolor.ts]:
```suggestion
  background?: boolean;
  name?: string;
  brightness?: number;
```

File: public/vendor/ansicolor/ansicolor.ts [hunk 6]
```diff
@@ -82,7 +82,7 @@ class Color {
     });
   }
 
-  defaultBrightness(value: number) {
+  defaultBrightness(value: number | undefined) {
```
Comment (dprokop) [inline public/vendor/ansicolor/ansicolor.ts]:
```suggestion
  defaultBrightness(value?: number) {
```

File: public/vendor/ansicolor/ansicolor.ts [hunk 7]
```diff
@@ -351,7 +351,7 @@ export default class Colors {
 
   get parsed() {
     let styles: Set<string>;
-    let brightness: number;
+    let brightness: number | undefined;
```
Comment (dprokop) [inline public/vendor/ansicolor/ansicolor.ts]:
```suggestion
    let brightness?: number;
```

File: public/vendor/ansicolor/ansicolor.ts [hunk 8]
```diff
@@ -62,7 +62,9 @@ const clean = (obj: any) => {
 class Color {
   background: boolean | undefined;
   name: string | undefined;
-  brightness: number | undefined;
+  background?: boolean;
+  name?: string;
+  brightness?: number;
```
Comment (dprokop) [inline public/vendor/ansicolor/ansicolor.ts]: you have duplicate properties now (background & name)

Comment (dehrax) [reply public/vendor/ansicolor/ansicolor.ts]: Fixed it yesterday, but forgot to push the changes

**Review by torkelo — APPROVED** [review]

**Pull request merged**